base/hashing.jl

```diff
@@ -64,11 +64,65 @@ end
 
 hash(x::QuoteNode, h::UInt) = hash(x.value, hash(QuoteNode, h))
 
-# hashing ranges by component at worst leads to collisions for very similar ranges
-const hashr_seed = UInt === UInt64 ? 0x80707b6821b70087 : 0x21b70087
+## hashing collections ##
+
+const hashaa_seed = UInt === UInt64 ? 0x7f53e68ceb575e76 : 0xeb575e76
+const hashrle_seed = UInt == UInt64 ? 0x2aab8909bfea414c : 0xbfea414c
+function hash(a::AbstractArray, h::UInt)
+    h += hashaa_seed
+    h += hash(size(a))
+
+    state = start(a)
+    done(a, state) && return h
+    x1, state = next(a, state)
+    # Always hash the first element
+    h = hash(x1, h)
+    done(a, state) && return h
+
+    # Then hash the difference between two subsequent elements when - is supported,
```

Comment:
You could have a fast-path version wrapped in `isleaftype(eltype(a))`, which gets elided at compile-time (see #7088). Would be lovely to have `applicable` work the same way, whenever all passed types are leaf-types.

Demo for `isleaftype`:
```jl
julia> function testelt{T}(a::Array{T})
           if isleaftype(T)
               return "leaf"
           end
           return "nonleaf"
       end
testelt (generic function with 1 method)

julia> @code_warntype testelt(Int[1])
Variables:
  #self#::#testelt
  a::Array{Int64,1}

Body:
  begin  # REPL[1], line 2:
      $(QuoteNode(true)) # REPL[1], line 3:
      return "leaf"
  end::String

julia> @code_warntype testelt(Integer[1])
Variables:
  #self#::#testelt
  a::Array{Integer,1}

Body:
  begin  # REPL[1], line 2:
      $(QuoteNode(false))
      goto 4
      4:  # REPL[1], line 5:
      return "nonleaf"
  end::String
```